### ISSUE TITLE
Barabasi aging game crashes on negative index from psumtree_search

### ISSUE THREAD

**Issue by GroteGnoom** [igraph/igraph#1618]
**Describe the bug**
A specific barabasi_aging_game call will crash.

**To reproduce**
```
igraph_rng_seed(igraph_rng_default(), 42);
assert(igraph_barabasi_aging_game(
        &g, /*nodes*/ 5, /*m: edges_per_step*/ 2, 
        /*outseq: edges per step as vector*/ 0, /*outpref*/ 0,
        /*pa_exp*/ 1.0, /*aging_exp*/ 0.0, /*aging_bin*/ 1,
        /*zero_deg_appeal*/ 1.0, /*zero_age_appeal*/ 1.0, /*deg_coef*/ -1.0,
        /*age_coef */ 0.0, /*directed*/ 1) == IGRAPH_SUCCESS);
```
This crashes with a double free/corruption on 
`igraph_vector_destroy(&degree);`
or if I remove that, it crashes on the igraph_create, because of negative indices:
If I print the edges I get this:
`1 0 1 0 2 1 2 1 3 2 3 2 4 -1 4 -1`
which is because
`igraph_psumtree_search(&sumtree, &to, RNG_UNIF(0, sum));`
sets values of `-1` to `to`.

This does not happen with less than 5 nodes, only 1 edge per step, or a deg_coef of 1.0  (if I keep the rest the same). The negative deg_coef is maybe a little weird, but I don't think it should be impossible.

**Version information**
a0d5abe7c84155c4d1a3885589037ed2470386a5, the current develop

https://github.com/igraph/igraph/blob/a0d5abe7c84155c4d1a3885589037ed2470386a5/src/games/barabasi.c#L602

**Comment by GroteGnoom** [comment]
Maybe this is a solution:
igraph_sumtree_update should return IGRAPH_EINVAL if new_value is negative.
igraph_barabasi_aging_game should use a IGRAPH_CHECK for the igraph_sumtree_update calls.

Additionaly, either igraph_sumtree_search should return IGRAPH_EINVAL when the search parameter is outside the range of zero to the root value of the tree, or it should always return -1 if the value is not found, which should be documented.

**Comment by ntamas** [comment]
I think that a negative `deg_coef` does not make sense because this would mean that the _probability_ of odd-degree nodes receiving a new edge would be negative, and I can't quite wrap my head around negative probabilities. (I'd appreciate a confirmation here from @szhorvat or @vtraag ). So I think we are dealing with two issues here:

1. `igraph_barabasi_aging_game` should disallow negative values for `deg_coef`. Most likely it also doesn't make sense to allow `deg_coef = age_coef = zero_deg_appeal = zero_age_appeal = 0` either because that would mean that all nodes have a zero probability of receiving a new edge.

2. `igraph_psumtree_update()` should deal with negative values.

First I want to check the code of `igraph_psumtree` in general to understand whether it makes sense to have negative values in there. Right now we use this data structure for drawing random numbers from a dynamic discrete distribution, so in this particular use-case it does not make sense to allow negative values, but there could be other use-cases that I don't see right now where negative values would make sense.

**Comment by szhorvat** [comment]
> I'd appreciate a confirmation here 

I agree.

----

Let's spell out how it works:

In the standard (non-aging) game, the probability of connecting to a vertex is

    weight = degree ^ pa_exp + zero_deg_appeal

The aging version modifies this to

    weight = deg_coef * (degree ^ pa_exp + zero_deg_appeal)  +  age_coef * (age ^ age_exp + zero_age_appeal)

Thus, to guarantee that this value is non-negative, `deg_coef` and `age_coef` must be non-negative as well.

`zero_deg_appeal` and `zero_age_appeal` must be non-negative (typically positive, but in some cases zero makes sense).

`pa_exp` and `age_exp` may be either negative or positive.

With these rules, the weight may still be zero for _all_ vertices even when there are multiple vertices, which is not good. However, this condition can arise under multiple complicated combinations of input parameters, so it's difficult to ensure it never happens. I would not try to check for it, and would leave it up to the user to ensure it doesn't happen.

**Comment by ntamas** [comment]
This is now fixed in 1d422e68c
